test/Conversion/ExportVerilog/emit-fragment.mlir

```diff
@@ -0,0 +1,66 @@
+// RUN: circt-opt -o=- --export-verilog %s | FileCheck %s
+// RUN: circt-opt -o=- --export-split-verilog='dir-name=%t' %s
+// RUN: cat %t%{fs-sep}FirstModule.sv | FileCheck %s --check-prefix=FIRST
+// RUN: cat %t%{fs-sep}SecondModule.sv | FileCheck %s --check-prefix=SECOND
+
+// CHECK:      `ifndef MacroA
+// CHECK-NEXT:   `define MacroA A
+// CHECK-NEXT: `endif
+// CHECK-NEXT: `ifndef MacroB
+// CHECK-NEXT:   `define MacroB B
+// CHECK-NEXT: `endif
+// CHECK-LABEL: FirstModule
+// CHECK-NOT: MacroB
+// CHECK:      `ifndef MacroC
+// CHECK-NEXT:   `define MacroC C
+// CHECK-NEXT: `endif
+// CHECK-LABEL: SecondModule
+
+// FIRST:      `ifndef MacroA
+// FIRST-NEXT:   `define MacroA A
+// FIRST-NEXT: `endif
+// FIRST-NEXT: `ifndef MacroB
+// FIRST-NEXT:   `define MacroB B
+// FIRST-NEXT: `endif
+// FIRST-NEXT: module FirstModule
+
+// SECOND:      `ifndef MacroB
+// SECOND-NEXT:   `define MacroB B
+// SECOND-NEXT: `endif
+// SECOND-NEXT: `ifndef MacroC
+// SECOND-NEXT:   `define MacroC C
+// SECOND-NEXT: `endif
+// SECOND-NEXT: module SecondModule
+
+sv.macro.decl @MacroA
+sv.macro.decl @MacroB
+sv.macro.decl @MacroC
+
+emit.fragment @FragmentA {
+  sv.ifdef @MacroA {
+  } else {
+    sv.macro.def @MacroA "A"
+  }
+}
+
+emit.fragment @FragmentB {
+  sv.ifdef @MacroB {
+  } else {
+    sv.macro.def @MacroB "B"
+  }
+}
+
+emit.fragment @FragmentC {
+  sv.ifdef @MacroC {
+  } else {
+    sv.macro.def @MacroC "C"
+  }
+}
+
+hw.module @FirstModule(in %in : i32, out out : i32) attributes { "emit.fragments" = [@FragmentA, @FragmentB] } {
```

Comment:
Isn't adding the required fragments as discardable attributes quite fragile?

I'm wondering if you considered also putting the `hw.module` in its own fragment that references `@FragmentA` and `@FragmentB` as dependencies. Then, a file can be constructed by choosing a set of fragments that should go there (e.g., because they contain modules that should be in that file). The exporter then includes all fragments it depends on transitively in a legal order.

(I don't know a lot about the goals of this, how the compiler decides which module(s) go in which file, and ExportVerilog in general. I'm looking at the emit dialect because in arcilator we'd like to also emit things in separate files for parallel LLVM compilation and using the emit dialect for that could be nice)

Reply:
`hw.module` will soon be referenced by `emit.ref` and we'll be able to remove this attribute, using refs instead.